sandbox/OpenDocument/docs/rst2odt.1
.TH RST2ODT "1" "January 2008" "docutils" "User Commands"
.SH NAME
rst2odt \- convert reST documents to ODT
.SH SYNOPSIS
.B rst2odt
[\fIoptions\fR] [\fIsource\fR [\fIdestination\fR]]
.SH DESCRIPTION
.PP
Generates OpenDocument Text (ODT) documents from reStructuredText sources.
Reads from \fIsource\fR (default is stdin) and writes to \fIdestination\fR
(default is stdout).
.SH OPTIONS

.SS
\fBGeneral Docutils Options\fR
.TP
\fB\-\-title\fR=\fITITLE\fR
Specify the document title as metadata (not part of
the document body).  Overrides a document-provided
title.  There is no default.
.TP
\fB\-\-generator\fR, \fB\-g\fR
Include a "Generated by Docutils" credit and link at
the end of the document.
.TP
\fB\-\-no\-generator\fR
Do not include a generator credit.
.TP
\fB\-\-date\fR, \fB\-d\fR
Include the date at the end of the document (UTC).
.TP
\fB\-\-time\fR, \fB\-t\fR
Include the time & date at the end of the document
(UTC).
.TP
\fB\-\-no\-datestamp\fR
Do not include a datestamp of any kind.
.TP
\fB\-\-source\-link\fR, \fB\-s\fR
Include a "View document source" link (relative to
destination).
.TP
\fB\-\-source\-url\fR=\fI<URL>\fR
Use the supplied <URL> verbatim for a "View document
source" link; implies \-\-source-link.
.TP
\fB\-\-no\-source\-link\fR
Do not include a "View document source" link.
.TP
\fB\-\-toc\-entry\-backlinks\fR
Enable backlinks from section headers to table of
contents entries.  This is the default.
.TP
\fB\-\-toc\-top\-backlinks\fR
Enable backlinks from section headers to the top of
the table of contents.
.TP
\fB\-\-no\-toc\-backlinks\fR
Disable backlinks to the table of contents.
.TP
\fB\-\-footnote\-backlinks\fR
Enable backlinks from footnotes and citations to their
references.  This is the default.
.TP
\fB\-\-no\-footnote\-backlinks\fR
Disable backlinks from footnotes and citations.
.TP
\fB\-\-section\-numbering\fR
Enable Docutils section numbering (default: enabled).
.TP
\fB\-\-no\-section\-numbering\fR
Disable Docutils section numbering (default: enabled).
.TP
\fB\-\-strip\-comments\fR
Remove comment elements from the document tree
(default: leave them).
.TP
\fB\-\-leave\-comments\fR
Leave comment elements in the document tree (this is
the default).
.TP
\fB\-\-report\fR=\fI<level>\fR, \fB\-r\fR \fI<level>\fR
Set verbosity threshold; report system messages at or
higher than <level> (by name or number: "info" or "1",
warning/2, error/3, severe/4; also, "none" or "5").
Default is 2 (warning).
.TP
\fB\-\-verbose\fR, \fB\-v\fR
Report all system messages, info-level and higher.
(Same as "\-\-report=info".)
.TP
\fB\-\-quiet\fR, \fB\-q\fR
Do not report any system messages.  (Same as
"\-\-report=none".)
.TP
\fB\-\-halt\fR=\fI<level>\fR
Set the threshold (<level>) at or above which system
messages are converted to exceptions, halting
execution immediately by exiting (or propagating the
exception if \-\-traceback set).  Levels as in \-\-report.
Default is 4 (severe).
.TP
\fB\-\-strict\fR
Same as "\-\-halt=info": halt processing at the
slightest problem.
.TP
\fB\-\-exit\-status\fR=\fI<level>\fR
Enable a non-zero exit status for normal exit if
non-halting system messages (at or above <level>) were
generated.  Levels as in \-\-report.  Default is 5
(disabled).  Exit status is the maximum system message
level plus 10 (11 for INFO, etc.).
.TP
\fB\-\-debug\fR
Report debug-level system messages and generate
diagnostic output.
.TP
\fB\-\-no\-debug\fR
Do not report debug-level system messages or generate
diagnostic output.
.TP
\fB\-\-warnings\fR=\fI<file>\fR
Send the output of system messages (warnings) to
<file>.
.TP
\fB\-\-traceback\fR
Enable Python tracebacks when halt-level system
messages and other exceptions occur.  Useful for
debugging, and essential for issue reports.
.TP
\fB\-\-no\-traceback\fR
Disable Python tracebacks when errors occur; report
just the error instead.  This is the default.
.TP
\fB\-\-input\-encoding\fR=\fI<name[:handler]>\fR, \fB\-i\fR \fI<name[:handler]>\fR
Specify the encoding of input text.  Default is
locale-dependent.  Optionally also specify the error
handler for undecodable characters, after a colon
(":"); default is "strict".
(See "\-\-input-encoding-error-handler".)
.TP
\fB\-\-input\-encoding\-error\-handler\fR=\fIINPUT_ENCODING_ERROR_HANDLER\fR
Specify the error handler for undecodable characters
in the input.  Acceptable values include "strict",
"ignore", and "replace".  Default is "strict".
Usually specified as part of \-\-input-encoding.
.TP
\fB\-\-output\-encoding\fR=\fI<name[:handler]>\fR, \fB\-o\fR \fI<name[:handler]>\fR
Specify the text encoding for output.  Default is
UTF-8.  Optionally also specify the error handler for
unencodable characters, after a colon (":"); default
is "strict".  (See "\-\-output-encoding-error-handler".)
.TP
\fB\-\-output\-encoding\-error\-handler\fR=\fIOUTPUT_ENCODING_ERROR_HANDLER\fR
Specify the error handler for unencodable characters
in the output.  Acceptable values include "strict",
"ignore", "replace", "xmlcharrefreplace", and
"backslashreplace" (in Python 2.3+).  Default is
"strict".  Usually specified as part of
\-\-output-encoding.
.TP
\fB\-\-error\-encoding\fR=\fI<name[:handler]>\fR, \fB\-e\fR \fI<name[:handler]>\fR
Specify the text encoding for error output.  Default
is ASCII.  Optionally also specify the error handler
for unencodable characters, after a colon (":");
default is "backslashreplace".
(See "\-\-output-encoding-error-handler".)
.TP
\fB\-\-error\-encoding\-error\-handler\fR=\fIERROR_ENCODING_ERROR_HANDLER\fR
Specify the error handler for unencodable characters
in error output.  See \-\-output-encoding-error-handler
for acceptable values.  Default is "backslashreplace".
Usually specified as part of \-\-error-encoding.
.TP
\fB\-\-language\fR=\fI<name>\fR, \fB\-l\fR \fI<name>\fR
Specify the language of input text (ISO 639 2-letter
identifier).  Default is "en" (English).
.TP
\fB\-\-record\-dependencies\fR=\fI<file>\fR
Write dependencies (caused e.g. by file inclusions) to
<file>.  Useful in conjunction with programs like
"make".
.TP
\fB\-\-config\fR=\fI<file>\fR
Read configuration settings from <file>, if it exists.
.TP
\fB\-\-version\fR, \fB\-V\fR
Show this program's version number and exit.
.TP
\fB\-\-help\fR, \fB\-h\fR
Show this help message and exit.

.SS
\fBreStructuredText Parser Options\fR
.TP
\fB\-\-pep\-references\fR
Recognize and link to standalone PEP references (like
"PEP 258").
.TP
\fB\-\-pep\-base\-url\fR=\fI<URL>\fR
Base URL for PEP references (default
"http://www.python.org/peps/").
.TP
\fB\-\-rfc\-references\fR
Recognize and link to standalone RFC references (like
"RFC 822").
.TP
\fB\-\-rfc\-base\-url\fR=\fI<URL>\fR
Base URL for RFC references (default
"http://www.faqs.org/rfcs/").
.TP
\fB\-\-tab\-width\fR=\fI<width>\fR
Set number of spaces for tab expansion (default 8).
.TP
\fB\-\-trim\-footnote\-reference\-space\fR
Remove spaces before footnote references.
.TP
\fB\-\-leave\-footnote\-reference\-space\fR
Leave spaces before footnote references.
.TP
\fB\-\-no\-file\-insertion\fR
Disable directives that insert the contents of
external file ("include" & "raw"); replaced with a
"warning" system message.
.TP
\fB\-\-file\-insertion\-enabled\fR
Enable directives that insert the contents of external
file ("include" & "raw").  Enabled by default.
.TP
\fB\-\-no\-raw\fR
Disable the "raw" directives; replaced with a
"warning" system message.
.TP
\fB\-\-raw\-enabled\fR
Enable the "raw" directive.  Enabled by default.

.SS
\fBStandalone Reader\fR
.TP
\fB\-\-no\-doc\-title\fR
Disable the promotion of a lone top-level section
title to document title (and subsequent section title
to document subtitle promotion; enabled by default).
.TP
\fB\-\-no\-doc\-info\fR
Disable the bibliographic field list transform
(enabled by default).
.TP
\fB\-\-section\-subtitles\fR
Activate the promotion of lone subsection titles to
section subtitles (disabled by default).
.TP
\fB\-\-no\-section\-subtitles\fR
Deactivate the promotion of lone subsection titles.

.SS
\fBODF-Specific Options\fR
.TP
\fB\-\-stylesheet\fR=\fI<URL>\fR
Specify a stylesheet URL, used verbatim.  Overrides
\-\-stylesheet-path.
.TP
\fB\-\-stylesheet\-path\fR=\fI<file>\fR
Specify a stylesheet file, relative to the current
working directory.  The path is adjusted relative to
the output ODF file.  Overrides \-\-stylesheet.
.TP
\fB\-\-cloak\-email\-addresses\fR
Obfuscate email addresses to confuse harvesters while
still keeping email links usable with
standards-compliant browsers.
.TP
\fB\-\-table\-border\-thickness\fR=\fITABLE_BORDER_THICKNESS\fR
Specify the thickness of table borders in thousands of
a cm.  Default is 35.
.TP
\fB\-\-add\-syntax\-highlighting\fR
Add syntax highlighting in literal code blocks.Default
is No.  Requires installation of Pygments.
.SH AUTHOR
This man page was generated from the \-\-help output of the tool it documents, using a script written by Simon McVittie for the Debian GNU/Linux system. The script may be used by others: please see the Debian source package python-docutils if you're interested.


prest/ChangeLog
.. -*-rst-*-

Release Notes
===============

.. contents::

Release 0.3.37
--------------

reStructuredText parsing:

* Added "-D relax-citation-characters" option.
* Added roles "mathml-display" and "mathml-inline" to force
  display/inline mode.
* Role name is now passed in as fourth argument to role attribute
  subroutines.
* DefineRole now allows defining other info about the role than just
  its options.
* Plug-in directives can now have an init routine that gets called to
  define and initialize them.
* Plug-ins can now define roles as well as directives.
* Roles can now define a reparse option which will cause the output to
  be reparsed as reStructuredText.

Writer 'latex':

* Added "-W chapter" and "-W cloak-email-addresses" options.
* Added a new '\styleliteralblock' style command.
* Fixed a problem with inferring minus signs.
* Fixed a problem with converting opening single quotes.
* Fixed handling of raw latex.


Release 0.3.36
--------------

Directive 'code-block':

* Allow the 'states' executable to be found anywhere on the path, but
  check standard locations first.

Writer 'latex':

* Sort the dependent \usepackage modules so that there is a canonical
  order for regression testing

Writer 'index':

* Sort the inline targets so that there is a canonical order for
  regression testing

Release 0.3.35
--------------

There were no functional changes in this release.

Fixed the latex simple test not to use a code-block directive, which
compiles differently on different machines.

Release 0.3.34
--------------

reStructuredText transforms:

* Fixed a bug with hyperlink resolution when a chained target went to
  an indirect target to an internal target.

Writer 'index':

* Fixed a bug in dealing with raw DOM objects.

Writer 'latex':

* Fixed bugs in line blocks
* Fixed bugs in parsed literals
* Fixed a bug where a paragraph ending in an interpreted text item did
  not result in a new paragraph
* Fixed a bug with processing quotes around marked up text
* Fixed a bug with special characters in options
* Added support for outputting mathml to latex (requires
  Text::ASCIIMathML 0.4 or later)

Release 0.3.33
--------------

There were no functional changes in this release.

* Fixed some tests and expects to be compatible with perl 5.9.5.
* Modified Makefile.PL hopefully to be compatible with more versions
  of make.
* Changed the way the system plug-in directive gets the message from 
  STDERR so that it will hopefully be compatible with more systems.

Release 0.3.32
--------------
prest script:

* Generated a fatal error if an input file cannot be opened.

ReStructuredText parsing:

* The target role now keeps the case of the target in creating its
  name.

Writer 'html':

* Process the :class: option on images.
* If the "-W image-exts" option is used (it's actually deprecated in
  favor of "-D image-exts"), the default alt attribute now also uses
  the substituted extension.

Writer 'index':

* Improved help text.
* Added "-W index-indirect-targets" option.
* Added "-W index-role-target" option.

Writer 'xref':

* Added a "-W xref-role-target" option to specify which target role
  (if any) should be used for cross references.
* The "-W xref-role-targets" option has been removed.

In general:

* Recoded to remove most dependencies on internal structure of DOM
  objects outside of DOM.pm.

Release 0.3.31
--------------
ReStructuredText parsing:

* Added a "target" role that allows separation of inline target tags
  from the text that is displayed.
* Fixed a bug where, if two items in an enumerated list occurred with
  no blank line between them and a third item was separated by a blank
  line, the third item was put into a separate list.

Directive 'perl':

* Exported the following new global variables: ``$DIRECTIVE`` and
  ``$PARSER``.

Writer 'html':

* Mark a single paragraph in a compound as both class "compound-first"
  and "compound-last".
* Fixed a bug where a class attribute on a field list did not
  propagate to the output.
* Fixed a bug where an align attribute on a table did not propagate to
  the output.

Writer 'index':

* Implemented the '-W doc-titles' and '-W index-role-targets' options.
* Output nothing instead of an empty table if there are no index
  entries.

Writer 'xref':

* Added the ``-W xref-role-targets`` option.
* Allowed the suffix to be an empty string.

Release 0.3.30
--------------
* Fixed problem with find command in Makefile.PL that used a Linuxism.
* Made most of the perl modules use the subversion release as version number.

Release 0.3.29
--------------
* Renamed ReleaseNotes.txt to ChangeLog so that it will appear as a
  special file on CPAN.
* Made one more effort to fix the indexing of version numbers.

Release 0.3.28
--------------
* Create lib/Text/Restructured.pm from lib/Text/Restructured.pm.PL so
  we can insert the version number in to allow CPAN to index it
  correctly.
* Fixed a problem with creating regular-expression expects for tests.

Release 0.3.27
--------------
Fixed problem with CPAN indexing of versions.

Release 0.3.26
--------------
* Completely reworked the regression test structure so as not to be
  dependent upon GNU make.

Writers 'dom' and 'xml':

* Fixed a bug with attribute strings that contain double quotes.


Release 0.3.25
--------------
* Fixed a problem with Makefile.PL.

ReStructuredText parsing:

* Don't write messages to STDERR for errors in directives in test
  data.

Writer 'html':

* Fixed a problem with quote markes in "alt" image text.
* Added "-W body-only" flag.

Release 0.3.24
--------------
* Fixed problems with running with -T.
* Fixed Makefile.

Release 0.3.23
--------------
* Fixed the help to parse as reStructuredText.
* Execute evals in ``Text::Restructured::Writer::Eval`` instead of ``Eval_::``.
* Initialize some of the opt variables to avoid warnings if not driven
  by prest.

ReStructuredText parsing:

* Added "-D image-exts" flag so that each writer doesn't have to

sandbox/OpenDocument/tools/rst2odt.py
#!/usr/bin/env python

# $Id$
# Author: Dave Kuhlman <dkuhlman@rexx.com>
# Copyright: This module has been placed in the public domain.

"""
A front end to the Docutils Publisher, producing OpenOffice documents.
"""

import sys
try:
    import locale
    locale.setlocale(locale.LC_ALL, '')
except:
    pass

from docutils.core import publish_cmdline, default_description, \
    Publisher, default_usage
from docutils import io
from docutils.writers.odtwriter import Writer


description = ('Generates OpenDocument/OpenOffice/ODF documents from '
               'standalone reStructuredText sources.  ' + default_description)


class BinaryFileOutput(io.FileOutput):
    """
    A version of docutils.io.FileOutput which writes to a binary file.
    """
    def open(self):
        try:
            self.destination = open(self.destination_path, 'wb')
        except IOError, error:
            if not self.handle_io_errors:
                raise
            print >>sys.stderr, '%s: %s' % (error.__class__.__name__,
                                            error)
            print >>sys.stderr, ('Unable to open destination file for writing '
                                 '(%r).  Exiting.' % self.destination_path)
            sys.exit(1)
        self.opened = 1


def publish_cmdline_to_binary(reader=None, reader_name='standalone',
                    parser=None, parser_name='restructuredtext',
                    writer=None, writer_name='pseudoxml',
                    settings=None, settings_spec=None,
                    settings_overrides=None, config_section=None,
                    enable_exit_status=1, argv=None,
                    usage=default_usage, description=default_description,
                    destination=None, destination_class=BinaryFileOutput
                    ):
    """
    Set up & run a `Publisher` for command-line-based file I/O (input and
    output file paths taken automatically from the command line).  Return the
    encoded string output also.

    This is just like publish_cmdline, except that it uses
    io.BinaryFileOutput instead of io.FileOutput.

    Parameters: see `publish_programmatically` for the remainder.

    - `argv`: Command-line argument list to use instead of ``sys.argv[1:]``.
    - `usage`: Usage string, output if there's a problem parsing the command
      line.
    - `description`: Program description, output for the "--help" option
      (along with command-line option descriptions).
    """
    pub = Publisher(reader, parser, writer, settings=settings,
        destination_class=destination_class)
    pub.set_components(reader_name, parser_name, writer_name)
    output = pub.publish(
        argv, usage, description, settings_spec, settings_overrides,
        config_section=config_section, enable_exit_status=enable_exit_status)
    return output


writer = Writer()
output = publish_cmdline_to_binary(writer=writer, description=description)
    
